### PR TITLE
Middlewares auth

### PR DESCRIPTION
**Pull request by mkkasem**
I created the middlewares but did not add them to the routes files yet

### CODE REVIEW

File: src/middlewares/item-authorization.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable no-underscore-dangle */
+const Item = require('../models/item');
+
+module.exports = async (req, res, next) => {
+  const itemId = req.params.id;
+  const item = await Item.findById(itemId);
+  if (item && item.owner._id.toString() === req.user._id) {
```
Comment (AmmarBaki2) [inline src/middlewares/item-authorization.js:7]: Since the `item.owner` is the reference to the user you don't need to use `._id`, 
`item.owner.toString()` is enough.

File: src/middlewares/user-authorization.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// this middlewawre is called after the the isAuth middleware
+// and checks if the user is authorized to access the route
+module.exports = function (req, res, next) {
+  // eslint-disable-next-line no-underscore-dangle
+  if (req.user._id === req.params.id) {
+    next();
+  } else {
```
Comment (AmmarBaki2) [inline src/middlewares/user-authorization.js:7]: Good job 👏 
We can use this when a user wants to delete his account

**Review by AmmarBaki2 — CHANGES_REQUESTED** [review]
Excellent job @mkkasem 👏 
There is a slight change I requested in the comments below

**Review by AmmarBaki2 — APPROVED** [review]
Well done 😀

**Review by mehmetfatiherdem — APPROVED** [review]

**Comment by mkkasem** [comment]
@Shrreya  can you please have a look at my code when you are free 😁?

**Comment by Shrreya** [comment]
> @Shrreya can you please have a look at my code when you are free 😁?

@mkkasem Are all the changes done? I had started reviewing but then I saw you were pushing few commits so I thought I'll wait.

**Comment by mkkasem** [comment]
> > @Shrreya can you please have a look at my code when you are free 😁?
> 
> @mkkasem Are all the changes done? I had started reviewing but then I saw you were pushing few commits so I thought I'll wait.

Yes I think I am done, but I haven't add these functions to the routes because I am not sure if I should add them now or the next sprint. I was planning to ask you about this issue after you approve this PR.

**Review by Peri7at — APPROVED** [review]

File: src/middlewares/__tests__/isAuth.test.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
+const jwt = require('jsonwebtoken');
+const isAuth = require('../isAuth');
+
+const validToken = { tokenValue: 'validToken', expires: 'validTime' };
```
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:4]: Better to rename this to: `validTokenWithValidTime`

File: src/middlewares/__tests__/isAuth.test.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+const validTokenWithNoValidTime = {
+  tokenValue: 'invalidToken',
+  expires: 'validTime',
+};
+const invalidToken = {
+  tokenValue: 'invalidToken',
+  expires: 'validTime',
+};
```
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:12]: Both these tokens are the same. Properties of one need to be changed.
Also rename one to `validTokenWithInvalidTime` and other to `invalidTokenWithValidTime`.

File: src/middlewares/__tests__/isAuth.test.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+const invalidTokenWithInvalidTime = {
+  tokenValue: 'invalidToken',
+  expires: 'invalidTime',
+};
+const user = {
+  id: 1,
+  username: 'name',
+  email: 'someemail@gmail.com',
+};
```
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:21]: You can rename this to `mockUser` which explains that this will be used as mock value in tests.

File: src/middlewares/__tests__/isAuth.test.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+const next = jest.fn().mockReturnThis();
+
+const verify = jest.spyOn(jwt, 'verify');
+
+afterAll(() => {
+  jest.clearAllMocks();
+});
+describe('isAuth function middleware', () => {
+  it('Should verify when valid token and correct secret are passed', (done) => {
```
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:36]: You can rephrase this to `Should authenticate for valid token with valid expiry time`.

File: src/middlewares/__tests__/isAuth.test.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      },
+    };
+
+    isAuth(req, res, next);
+    expect(verify).toHaveBeenCalled();
+    expect(verify).toHaveBeenCalledTimes(1);
+    expect(res.json).not.toHaveBeenCalled();
+    expect(verify).not.toThrow();
+    expect(verify).toHaveBeenCalledWith(validToken, process.env.JWT_SECRET);
+    expect(verify).toHaveReturnedWith(user);
```
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:50]: You should also check that `next()` has been called and `req.user` value is set.

File: src/middlewares/__tests__/isAuth.test.js [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    done();
+  });
+
+  it('Should throw error when token time is expired', (done) => {
```
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:54]: You can rephrase this to `Should not authenticate for valid token with invalid expiry time`

File: src/middlewares/__tests__/isAuth.test.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  username: 'name',
+  email: 'someemail@gmail.com',
+};
+
+const res = {
+  status: jest.fn().mockReturnThis(),
+  json: jest.fn().mockReturnThis(),
+  clearCookie: jest.fn().mockReturnThis(),
+};
+const next = jest.fn().mockReturnThis();
```
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:28]: You should add `mockReturnThis()` only for functions that will be chained in your implementation.

File: src/middlewares/__tests__/isAuth.test.js (hunk 8 above)
Comment (Shrreya) [reply src/middlewares/__tests__/isAuth.test.js:54]: Check if other test statements also need to rephrased. Try to follow similar words throughout test statements.

File: src/middlewares/__tests__/item-authorization.test.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    return Promise.resolve(null);
+  }),
+}));
+
+afterAll(() => {
+  jest.clearAllMocks();
+});
+describe('itemAuthorization function middleware ', () => {
+  it('should authorize user when his id equals to item owner id', async () => {
```
Comment (Shrreya) [inline src/middlewares/__tests__/item-authorization.test.js:30]: Try to avoid mentioning implementation logic in test statements.
So here you can rephrase to `Should authorize when requesting user is the owner of requested item`.

Comment (Shrreya) [reply src/middlewares/__tests__/item-authorization.test.js:30]: Check the other test statements too and see if there's any rephrasing required similar to this.

**Review by Shrreya — CHANGES_REQUESTED** [review]
@mkkasem Good job with the implementation! Logic looks solid.

Some improvements related to best practices are required.

**Comment by Shrreya** [comment]
@mkkasem Can you also merge the main branch into your branch? To ensure latest changes from your team mates are included.

File: src/middlewares/isAuth.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
+const jwt = require('jsonwebtoken');
+
+module.exports = async (req, res, next) => {
+  const { token } = req.cookies;
+  try {
+    const { user } = jwt.verify(token, process.env.JWT_SECRET);
```
[the remark this answers is not in the document]
Comment (mkkasem) [reply src/middlewares/isAuth.js:6]: yes, I was getting the user without deconstruction, but it was returning undefined because the verify function returns an object that contains the user object along with some properties like expiration time that's why I used the deconstruction

File: src/middlewares/isAuth.js [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    // if verified pass the user to the next function with the req object
+    req.user = user;
+    return next();
```
[the remark this answers is not in the document]
Comment (Shrreya) [reply src/middlewares/isAuth.js:9]: I think you can fix that by removing `async` in front of the arrow function, since there's no `await` call inside.

[the remark this answers is not in the document]
Comment (mkkasem) [reply src/middlewares/isAuth.js:9]: I did that but still getting the same error.

File: src/middlewares/user-authorization.js [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+  try {
+    // eslint-disable-next-line no-underscore-dangle
+    if (req.user._id === req.params.id) {
+      return next();
+    }
+    return res.status(401).json({ message: 'Unauthorized' });
+  } catch (error) {
+    return res.status(401).json({ message: 'Unauthorized' });
+  }
+};
```
[the remark this answers is not in the document]
Comment (mkkasem) [reply src/middlewares/user-authorization.js:13]: Put  /api/users/:id
a user can only update his information 
also for deletion

File: package.json [hunk 14]
```diff
@@ -7,7 +7,7 @@
   "license": "MIT",
   "scripts": {
     "start": "nodemon src/app.js",
-    "test": "export IS_JEST=1 && yarn validate && jest --passWithNoTests",
+    "test": "set IS_JEST=1 && yarn validate && jest --passWithNoTests",
```
Comment (AmmarBaki2) [inline package.json]: @mkkasem if you look at the discussion between me and Shrreya on the #23 PR you will get a better view on this. 
`set` doesn't work in github, you need to use it locally but when you commit make sure that it is `export` not `set` 
Just when you are going to push them to github after committing you can revert it back to `set` so it works on you terminal

Comment (AmmarBaki2) [inline package.json]: @mkkasem if you look at the discussion between me and Shrreya on the https://github.com/RCDD-202110-TUR-BEW/backend-capstone-turkey-paying-it-forward/pull/23 PR you will get a better view on this.
set doesn't work in github, you need to use it locally but when you commit make sure that it is export not set
Just when you are going to push them to github after committing you can revert it back to set so it works on you terminal

**Review by AmmarBaki2 — CHANGES_REQUESTED** [review]

File: package.json (hunk 14 above)
Comment (mkkasem) [reply package.json]: I saw the conversation and I know that I need to set it back to export, which I did in the last commit because I wasn't able to commit , so I fixed it on github immediately but I didn't know that if I change it to export without commiting it could work. 
Thanks ammar

Comment (AmmarBaki2) [reply package.json]: You are welcome 😊 
I faced the same problem at first, then I noticed that when you are pushing (as you know it runs the test script), it uses the script that is present in your local repo despite what is present in the commit.

File: src/middlewares/isAuth.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (mehmetfatiherdem) [reply src/middlewares/isAuth.js]: I got the same error in another function. I solved it by adding a res.json()

File: src/middlewares/__tests__/user-authentication.test.js [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    expect(verify).toHaveBeenCalledTimes(1);
+    expect(res.json).not.toHaveBeenCalled();
+    expect(verify).not.toThrow();
+    expect(verify).toHaveBeenCalledWith(
+      validTokenWithValidTime,
+      process.env.JWT_SECRET
+    );
+    expect(next).toHaveBeenCalled();
+    expect(req.user).toEqual(mockUser);
+    expect(req.user.id).toEqual(mockUser.id);
```
Comment (Shrreya) [inline src/middlewares/__tests__/user-authentication.test.js:59]: Here as well try to keep the order of execution in the order of your expect statements.

So first check `req.user`, then do all the checks on `verify` and last on `next` and `res.json` or `res.status`.

Also always run `toHaveBeenCalled` first, then `toHaveBeenCalledTimes`, then `toHaveBeenCalledWith` and last `toHaveReturnedWith` or `toThrow`.

File: src/middlewares/__tests__/user-authorization.test.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+const res = {
+  status: jest.fn().mockReturnThis(),
+  json: jest.fn().mockReturnThis(),
+};
+
+afterAll(() => {
+  jest.resetAllMocks();
+});
+describe('user-authorization function middleware', () => {
+  it('Should authorize user when he tries to modify his information', () => {
```
Comment (Shrreya) [inline src/middlewares/__tests__/user-authorization.test.js:20]: Let's keep our language gender-neutral. So you can say `Should authorize user when trying to modify own information`.

File: src/middlewares/__tests__/user-authorization.test.js [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    expect(next).not.toHaveBeenCalled();
+    expect(res.status).toHaveBeenCalled();
+    expect(res.status).toHaveBeenCalledWith(401);
+    expect(res.json).toHaveBeenCalled();
+    expect(res.json).toHaveBeenCalledWith({
+      message: 'Unauthorized to to modify the requested user',
+    });
+  });
+
+  it('should unauthorize user when user is not defined', () => {
```
Comment (Shrreya) [inline src/middlewares/__tests__/user-authorization.test.js:55]: Better statement could be `Should unauthorize user when user ID is not found`

File: src/middlewares/__tests__/user-authorization.test.js [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        id: 'anyId',
+      },
+    };
+    const next = jest.fn();
+
+    userAuthorization(req, res, next);
+    expect(next).not.toHaveBeenCalled();
+    expect(res.status).toHaveBeenCalledWith(401);
+    expect(res.json).toHaveBeenCalledWith({
+      message: 'Unauthorized to to modify the requested user',
```
Comment (Shrreya) [inline src/middlewares/__tests__/user-authorization.test.js:67]: Seems like `to` added twice by mistake.

File: src/middlewares/item-authorization.js [hunk 19]
```diff
@@ -0,0 +1,22 @@
+/* this middleware is called after the the user-authentication middleware to verify
+ that user is allowed to access some item endpoints that need authorization to be accessed */
```
Comment (Shrreya) [inline src/middlewares/item-authorization.js:2]: You can shorten this as:
```
/* this middleware is called after the user-authentication middleware to verify
 that user is authorized to access some protected item endpoints */

File: src/middlewares/user-authentication.js [hunk 20]
```diff
@@ -0,0 +1,19 @@
+/* this middle is called in all routes except the global ones to authenticate user */
```
Comment (Shrreya) [inline src/middlewares/user-authentication.js:1]: It would be better to write this as:
```
/* this middleware is called on all routes except the global ones to verify that user is authenticated */

File: src/middlewares/user-authorization.js [hunk 21]
```diff
@@ -0,0 +1,18 @@
+/* this middleware is called after the the user-authentication middleware to verify
+ that user is allowed to access some user endpoints that need authorization to be accessed */
```
Comment (Shrreya) [inline src/middlewares/user-authorization.js:2]: Rewrite this similar to the comment above about item-authorization.

**Comment by mkkasem** [comment]
@Shrreya all changes are done and I merged the main branch also.
I hope it's ready to be merged now :)

**Comment by mkkasem** [comment]
@Ammar-64 Could you please merge this branch when you are free?

**Review by AmmarBaki2 — APPROVED** [review]
